packages/cli/src/lib/typeDistProject.ts

```diff
@@ -0,0 +1,278 @@
+            if (distTypeMap[moduleName]) {
+              resolvedModules.push({
+                resolvedFileName: distTypeMap[moduleName],
+                isExternalLibraryImport: false,
+                resolvedUsingTsExtension: false,
+              });
+              continue;
+            }
+
+            // Handle resolving builtin node modules to @types/node
+            if (moduleName.startsWith('node:')) {
+              moduleName = moduleName.slice(5);
+            }
+            if (builtinModules.includes(moduleName)) {
+              const result = ts.resolveModuleName(
+                `@types/node/${moduleName}`,
+                containingFile,
+                compilerOptions,
+                moduleResolutionHost,
+              );
+
+              if (result.resolvedModule) {
+                resolvedModules.push(result.resolvedModule);
+                continue;
+              }
+            }
+
+            // Handle resolving relative paths and external node_modules.
+            const result = ts.resolveModuleName(
+              moduleName,
+              containingFile,
+              compilerOptions,
+              moduleResolutionHost,
+            );
+
+            if (result.resolvedModule) {
+              resolvedModules.push(result.resolvedModule);
+              continue;
+            }
+
+            throw new Error(`Failed to resolve module: ${moduleName}`);
+          }
+
+          return resolvedModules;
+        },
+      };
+    },
+  });
+};
+
+// A list of the package roles we want to extract features for
+const targetPackageRoles: PackageRole[] = [
+  'backend-plugin',
+  'backend-plugin-module',
+  'frontend-plugin',
+  'frontend-plugin-module',
+  'web-library',
+  'node-library',
+];
+
+// A list of the feature types we want to extract from the project
+// and include in the metadata
+const targetFeatureTypes = [
+  '@backstage/BackendFeature',
+  '@backstage/BackstagePlugin',
+  '@backstage/FrontendPlugin',
+  '@backstage/FrontendModule',
+] as const;
+
+export type BackstagePackageFeatureType = (typeof targetFeatureTypes)[number];
+
+export const getEntryPointDefaultFeatureType = (
+  role: PackageRole,
+  packageDir: string,
+  project: Project,
+  entryPoint: EntryPoint,
+): BackstagePackageFeatureType | null => {
+  if (isTargetPackageRole(role)) {
+    const dtsPath = resolvePath(
+      getDistTypeRoot(packageDir),
+      entryPoint.path.replace(entryPoint.ext, '.d.ts'),
+    );
+
+    try {
+      const defaultFeatureType = getSourceFileDefaultFeatureType(
+        project.addSourceFileAtPath(dtsPath),
+      );
+
+      if (defaultFeatureType) {
+        return defaultFeatureType;
+      }
+    } catch (error) {
+      console.error(
+        `Failed to extract default feature type from ${dtsPath}, ${error}. ` +
+          'Your package will publish fine but it may be missing metadata about its default feature.',
+      );
```

Comment:
Let's be strict and enforce proper metadata in packages published with the CLI for now. If it turns out that there are some edge cases that breaks this then we could provide some escape hatch in the future, but for now I'd prefer to ensure that this metadata exists for all packages.

Reply:
Never mind this, I think we can start by being a bit lenient here since the intention is for this to always be added in the end automatically anyway